File: app/spotify/spotify.py
import os
import time
import requests
import logging
import base64
from urllib.parse import urlencode
from db.database import insert_playback_data, get_spotify_creds, set_spotify_creds


logging.basicConfig(level=logging.INFO)

CLIENT_ID = os.getenv("CLIENT_ID")
CLIENT_SECRET = os.getenv("CLIENT_SECRET")
REDIRECT_URI = "http://localhost:5001/setup" # TODO: Change this
SCOPE = "user-read-currently-playing"

class SpotifyController:
    def __init__(self):
        self.is_authenticated = False
        self.auth_code     = ""
        self.access_token  = ""
        self.refresh_token = ""
        self.expiry        = 0

        auth_data = get_spotify_creds()
        if auth_data:
            self.is_authenticated = True # Will refresh if it needs it
            self.access_token     = auth_data['access_token']
            self.refresh_token    = auth_data['refresh_token']
            self.expiry           = auth_data['expiry']

    def __repr__(self):
        return f"Spotify(auth={self.is_authenticated},expiry={self.expiry})"
 
    def get_auth_redirect_url(self):
        params = { 
            "client_id": CLIENT_ID,
            "response_type": "code",
            "redirect_uri": REDIRECT_URI,
            "scope": SCOPE
        }
        return f"https://accounts.spotify.com/authorize?{urlencode(params)}"

    def set_auth_data(self, access, refresh, expiry):
        self.access_token = access
        self.refresh_token = refresh
        self.expiry = int(expiry)
        insert_ret = set_spotify_creds(access, refresh, int(expiry))
        self.is_authenticated = bool(insert_ret)
        if not insert_ret:
            self.is_authenticated = False
            logging.error("Error inserting auth data to database!")

    def has_token_expired(self):
        return time.time() > self.expiry

    def authenticate(self, refresh = False, code = None):
        if code:
            self.auth_code = code
        
        headers = { "Content-Type": "application/x-www-form-urlencoded" }
        payload = {}
        if refresh and self.refresh_token:
            auth_header = base64.b64encode(f"{CLIENT_ID}:{CLIENT_SECRET}".encode()).decode()
            headers["Authorization"] = f"Basic {auth_header}"

            payload = {
                "grant_type": "refresh_token",
                "refresh_token": self.refresh_token,
            }
        else:
            payload = { 
                "grant_type": "authorization_code", 
                "code": self.auth_code,
                "redirect_uri": REDIRECT_URI, 
                "client_id": CLIENT_ID, 
                "client_secret": CLIENT_SECRET 
            }
        
        logging.info(f"Using {payload}")
        response = requests.post("https://accounts.spotify.com/api/token", headers=headers, data=payload)
        try:
            response.raise_for_status()
            data = response.json()

            self.set_auth_data(
                data['access_token'], 
                data.get('refresh_token', self.refresh_token), # backup to existing token if spotify doesnt give a new one
                time.time() + int(data['expires_in'])
            )
        except Exception as e:
            logging.error(f"Error authenticating Spotify API: {e}")
            self.set_auth_data("", "", 0)
            self.is_authenticated = False

    def get_now_playing(self):
        if not self.is_authenticated:
            logging.info("Not authenticated: Visit http://localhost/ to authenticate")
            return
        if self.has_token_expired():
            logging.info("Reauthenticating")
            self.authenticate(refresh=True)
            return

        headers = { "Authorization": f"Bearer {self.access_token}"}
        response = requests.get("https://api.spotify.com/v1/me/player/currently-playing", headers=headers)
        if response.status_code == 204:
            logging.info("No song playing")
            return
        
        response.raise_for_status()
        data = response.json()
        try:
            insert_playback_data(data)
        except Exception as e:
            logging.error(e)

        

File: server/spotify/spotify.go
package spotify

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const (
	authorizeURL = "https://accounts.spotify.com/authorize"
	tokenURL     = "https://accounts.spotify.com/api/token"
	scopes       = "user-read-currently-playing user-read-playback-state user-read-recently-played"
)

type TokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresIn    int64  `json:"expires_in"`
	Scope        string `json:"scope"`
}

func AuthorizeURL(clientID, redirectURI, state string) string {
	params := url.Values{
		"client_id":     {clientID},
		"response_type": {"code"},
		"redirect_uri":  {redirectURI},
		"scope":         {scopes},
		"state":         {state},
	}
	return authorizeURL + "?" + params.Encode()
}

func ExchangeCode(clientID, clientSecret, code, redirectURI string) (*TokenResponse, error) {
	data := url.Values{
		"grant_type":   {"authorization_code"},
		"code":         {code},
		"redirect_uri": {redirectURI},
	}

	req, err := http.NewRequest("POST", tokenURL, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString(
		[]byte(clientID+":"+clientSecret),
	))

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("spotify token exchange failed: %s", resp.Status)
	}

	var token TokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&token); err != nil {
		return nil, err
	}
	return &token, nil
}

func RefreshToken(clientID, clientSecret, refreshToken string) (*TokenResponse, error) {
	data := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {refreshToken},
	}

	req, err := http.NewRequest("POST", tokenURL, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString(
		[]byte(clientID+":"+clientSecret),
	))

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token refresh failed: %s", resp.Status)
	}

	var token TokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&token); err != nil {
		return nil, err
	}
	return &token, nil
}

type PlaybackState struct {
	IsPlaying            bool     `json:"is_playing"`
	CurrentlyPlayingType string   `json:"currently_playing_type"`
	Item                 *Track   `json:"item"`
	ProgressMs           int      `json:"progress_ms"`
	Device               Device   `json:"device"`
	ShuffleState         bool     `json:"shuffle_state"`
	RepeatState          string   `json:"repeat_state"`
	Context              *Context `json:"context"`
}

type Track struct {
	ID         string   `json:"id"`
	Name       string   `json:"name"`
	Album      Album    `json:"album"`
	Artists    []Artist `json:"artists"`
	DurationMs int      `json:"duration_ms"`
	Popularity int      `json:"popularity"`
	IsLocal    bool     `json:"is_local"`
}

type Album struct {
	Name string `json:"name"`
}

type Artist struct {
	Name string `json:"name"`
}

type Device struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

type Context struct {
	URI string `json:"uri"`
}

func GetPlaybackState(accessToken string) (*PlaybackState, error) {
	req, err := http.NewRequest("GET", "https://api.spotify.com/v1/me/player", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNoContent {
		return nil, nil
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("playback state request failed: %s", resp.Status)
	}

	var ps PlaybackState
	if err := json.NewDecoder(resp.Body).Decode(&ps); err != nil {
		return nil, err
	}
	return &ps, nil
}

func ExpiryFromNow(expiresIn int64) int64 {
	return time.Now().Unix() + expiresIn
}
